Feature: Minimize window(s) instead of only switching to previous active window

hc42 opened the pull request:
Based on my own need I picked up  the idea of #12 . 
I have added better multi window handling.
I felt free to rearrange the code a little and added an editorconfig file describing the two space alignment.

What do you think?

adrienverge reviewed (commented):
Hello Hans-Christian, thanks for the proposal! 

I like the idea of minimizing the window instead of switching to the last previous one, and why not handling multiple windows if it's easy to do (in my use case, I've never used more than one).

The code looks nice :+1: although I couldn't really check it nor understand it, because this commit mixes multiples changes. Could you please split in multiple logical commits, e.g. refactors vs. feature changes? Each with a brief explanation of why/how in the commit message, otherwise I cannot review without spending too much time (which I lack). Also you can remove files `.editorconfig` and `.gitignore`, that are personal tastes.

Commit:
refactor: Global variables, variable names and early return

* Define global class variables on top and set to null if not used rather than deleting and adding:
  + Make the widget state easier visible
  + use null indicate an unset/empty state (whereas undefined indicates the variable does not exist)
* Added braces in for loop
  + technically possible without, but more error-prone and makes the reader harder think about it
* Introduced early return in the `magic_key_pressed` function (this might be a matter of style and taste)
  + Technically no difference, but reduces the mental load while reading the code. Else-If requires the reader to think about the previous if harder.
  + Allows to separate blocks of subsequent tasks, that can each be explained, so the reader can faster pick up what's happening.
* Rename the variable `magic` to `next` in `magic_key_pressed`
  + The meaning of magic is a little unclear here. Since `current` refers to the current window the name `next` makes more sense (for example in the if statement where the IDs are compared).

Commit:
refactor: Window getters aligned

* get_active_window uses global.display.focus_window
  + Use global variable instead of implicit index
  + Wrap in same object structure as the get_windows function provides
* get_windows wrapper object refactoring
  + active_window now returns a meta_window instance therefore use here also the meta_window instance to have all window objects of the same type. This also facilitates the handling in magic_key_pressed as this function operates only on the meta_window object.

hc42 commented on README.md:
```diff
@@ -5,7 +5,9 @@ Bind keys to a specific program in Gnome Shell:
-  front, the last window is brought to the front again.
+  front, the next window of the same programm is brought to the front.
+* When the corresponding key is pressed and the last window of this program
+  is already at the front, all windows of the programm will be minimized.
```
Thanks, if have applied that during PR refactoring.

hc42 commented:
Hi Adrien,
I finally found some time to get back to this. Sorry for letting you wait so long.
If added some refactoring commits with explanation of my ideas in the commit message. I hope that helps to understand my ideas. Please let me know if you find something more to adjust.

adrienverge commented:
Hello Hans-Christian,

Thanks, the 3 commit message are crystal-clear, and it's easier to review/understand with the split in 3 independent parts :+1:

I just noticed one small bug: if the magic key is pressed just after opening the Gnome Shell session, while no windows are open, it doesn't work. This bug does not exist on master branch. I guess it's related to `get_active_window()`. Could you fix that?

hc42 commented:
Hm, I can't reproduce this problem. However by looking at the code, the problem might be that the `global.display.focus_window` is null in your case.  Then accessing `.get_wm_class()` will cause an error.
I found [no clear documentation](https://mutter.gnome.org/meta/property.Display.focus-window.html) about the possible states, however looking at the [source code](https://github.com/GNOME/mutter/blob/main/src/core/display.c#L963) I found a few cases where `focus_window` is set to null.

 I have added a check in the `get_active_window()` function for that case. Can you try if it fixes the problem?
If not so can you give me some hint about your setup to reproduce? In my case Ubuntu 24.04.1 with Gnome 46 under Wayland.

adrienverge commented:
> Hm, I can't reproduce this problem. However by looking at the code, the problem might be that the `global.display.focus_window` is null in your case. Then accessing `.get_wm_class()` will cause an error. I found [no clear documentation](https://mutter.gnome.org/meta/property.Display.focus-window.html) about the possible states, however looking at the [source code](https://github.com/GNOME/mutter/blob/main/src/core/display.c#L963) I found a few cases where `focus_window` is set to null.

To reproduce the problem, I really have to log out then log in a new Gnome Shell session (no window must have been opened). Anyway, you found the cause because your latest `if (!current) return null;` did the job: it works perfectly now :tada: 

However I just noticed another problem (sorry for detecting it only now), when using [multiple workspaces](https://help.gnome.org/users/gnome-help/stable/shell-workspaces.html.en) (which I do, and other users probably do too). Before (on branch master), pressing the magic key would always show the window of the favorite program. With this PR's last commit, it's no longer the case if the favorite window is in another workspace. Worse: it gets minified/unminified in the background, without any visual feeback to the user.
Said differently: on master branch, pressing the key would switch to the correct worspace if needed.
I guess this is linked to minifying instead of bringing the last active window to the front. But it's probably solvable?

hc42 commented:
I finally found some time:
I guess you are using X11. It turns out that Wayland behaves different and does some things automatically X11 does not. So I didn't notice because I only tested on my setup "Works on my machine 🙃".
To fix this I added workspace detection in `get_active_window` (line 108) and active workspace switching in `﻿﻿focus_window`(line 128 ff). I also noticed that while toggeling through the windows the order always changes after minimizing. For me this is anoying. I enforced a constant order by sorting the windows (line 151 ff.).

adrienverge commented:
Hello, I use Wayland too :slightly_smiling_face:

But your latest push didn't solve the problem (at least on one of my laptops).
Steps to reproduce:
- Log out then log in a fresh new Gnome Shell session on Wayland.
- On workspace 1, launch your magic program (+ another regular window if you want).
- Go to workspace 2, press the magic key. Nothing happens (we stay on workspace 2, which is empty). In addition to this bug, notice that when manually going on workspace 1, your magic program got minimized (while it shouldn't be).

I took time to debug, and the cause seems to come from `if (current && current.title !== next.title)`. Changing it to `if (!current || current.title !== next.title)` solves the bug.
Does it make sense? (Please check because I haven't tested all use cases, especially with multiple magic windows!)
If yes, I propose to use it + to revert your latest push for X11 (no one should use Xorg nowadays).
```diff
@@ -105,7 +105,8 @@ export default class GnomeMagicWindowExtension extends Extension {
 
   get_active_window() {
     const current = global.display.focus_window;
+    // Revert latest @hc42's modification:
-    if (!current || !current.get_workspace().active) {
+    if (!current) {
       return null;
     }
     return {
@@ -125,10 +126,7 @@ export default class GnomeMagicWindowExtension extends Extension {
     }
     window.raise();
     Main.activateWindow(window);
+    // Revert latest @hc42's modification:
-    const workspace = window.get_workspace();
-    if (!workspace.active) {
-      workspace.activate(0);
-    }
   }
 
   hide_window(window) {
@@ -147,10 +145,16 @@ export default class GnomeMagicWindowExtension extends Extension {
     // log(this.debug());  // visible in journalctl -f
 
     const current = this.get_active_window();
+    // Revert latest @hc42's modification:
-    const magic_windows = this.find_magic_windows(title)
-      .sort((a, b) => a.ref.get_id() > b.ref.get_id()); // prevent flapping window order
-    const nextIndex = (magic_windows.findIndex(w => current && w.ref === current.ref) + 1) % magic_windows.length;
-    const next = magic_windows[nextIndex];
+    const magic_windows = this.find_magic_windows(title);
+    const next = magic_windows[0];
+
+    // Some helpful debugging lines:
+    Util.spawn(['/bin/bash', '-c', `echo 'magic_key_pressed()' >> /tmp/test`]);
+    Util.spawn(['/bin/bash', '-c', `echo '  current = ${JSON.stringify(current)}' >> /tmp/test`]);
+    Util.spawn(['/bin/bash', '-c', `echo '  magic_windows = ${JSON.stringify(magic_windows)}' >> /tmp/test`]);
+    Util.spawn(['/bin/bash', '-c', `echo '  next = ${JSON.stringify(next)}' >> /tmp/test`]);
+    Util.spawn(['/bin/bash', '-c', `echo '  this._first_window = ${this._first_window}' >> /tmp/test`]);
 
     // Launch application if no window of it found.
     if (!next) {
@@ -163,7 +167,7 @@ export default class GnomeMagicWindowExtension extends Extension {
     }
 
     // Focus "next" window, if other application is focused.
-    if (current && current.title !== next.title) {
+    if (!current || current.title !== next.title) {
       this._first_window = next.ref;
       this.focus_window(next.ref);
       return;
```

Even with this bug fixed, I found that switching windows:
\- works in one way: pressing the magic key now correctly shows the window of the favorite program, bringing me to the correct workspace if needed,
\- doesn't work in the other way: pressing the magic key again doesn't bring me to the workspace of the previously-active window.
I guess this is not a bug, since your proposal is to minimize the magic window instead of switching to previously-active one. Do you confirm it's intended? If yes, it's not my favorite behavior but why not, I think I can get used to it :slightly_smiling_face:

hc42 commented:
Hi,

thanks for taking the time and figuring this out with me. You spotted indeed the (logic) problem of the code. A problem that I didn't notice since my machines behave differently again 🙄. While trying to reproduce your machines behaviour I have tested Ubuntu 24.10 with Gnome 47 / Wayland. Works ok for me, so I added 47 to the version list.

If I remove the workspace focus code it does not work properly under some conditions on Gnome 47 for some reason 🙃. So I would prefer keeping it. Also the window order in the latest version makes more sense for me. However, this is in the end a matter of personal preference as the minimizing and focus to previous application behaviour. In deed the behaviour you described is my prefered option. I have about seven different dedicated magic keys (F13 -> F19) on my keyboard. I do not want to go back to an other application or workspace when pressing the dedicated application key. But in different usage scenarios, e.g., toggling a terminal window, this does make sense. Therefore I propose the following change: Config options.
I have added two options, resulting in four possible modes for the end of the window iteration:
A. Foucus only / endless cycling through the magic windows
B. Focus to the previous application only (old behaviour, set in this PR)
C. Minimizing the magic application windows (my prefered behaviour)
D. Minimizing the magic application windows and focus to the previous application
This way users can simply choose. 😀

Does this solution makes sense for you?

Commit:
Feature: Minimize window(s) instead of only switching to previous active window
